File: frontend/src/lib/permissions.ts
/**
 * Permissions par rôle — miroir de backend/app/core/permissions.py.
 *
 * Cette version ne fait PAS autorité : elle sert uniquement à l'affichage
 * (masquer un menu, un bouton, rediriger loin d'une page non autorisée).
 * L'application réelle des droits reste côté FastAPI (voir require_permission) —
 * un utilisateur qui contournerait cette couche resterait bloqué par l'API.
 */
'use client';

import { useEffect, useState } from 'react';

export type Module =
  | 'dashboard'
  | 'products'
  | 'categories'
  | 'stock'
  | 'orders'
  | 'returns'
  | 'clients'
  | 'debts'
  | 'suppliers'
  | 'finance'
  | 'team'
  | 'audit'
  | 'settings';

export type Role = 'owner' | 'manager' | 'stock_manager' | 'cashier' | 'staff' | 'admin';

const ALL_SELLING_ROLES: Role[] = ['owner', 'manager', 'stock_manager', 'cashier', 'staff'];

export const PERMISSIONS: Record<Module, Record<string, Role[]>> = {
  dashboard: {
    view: ['owner', 'manager'],
  },
  products: {
    view: ALL_SELLING_ROLES,
    write: ['owner', 'manager', 'stock_manager'],
    delete: ['owner', 'manager', 'stock_manager'],
  },
  categories: {
    view: ALL_SELLING_ROLES,
    write: ['owner', 'manager', 'stock_manager'],
    delete: ['owner', 'manager', 'stock_manager'],
  },
  stock: {
    write: ['owner', 'manager', 'stock_manager'],
  },
  orders: {
    view: ALL_SELLING_ROLES,
    create: ['owner', 'manager', 'cashier', 'staff'],
    cancel: ['owner', 'manager'],
  },
  returns: {
    create: ['owner', 'manager', 'stock_manager', 'cashier'],
  },
  clients: {
    view: ALL_SELLING_ROLES,
    create: ALL_SELLING_ROLES,
    edit: ['owner', 'manager', 'cashier'],
    delete: ['owner', 'manager'],
  },
  // Miroir exact de PERMISSIONS["debts"] dans core/permissions.py (Phase 4).
  debts: {
    view: ALL_SELLING_ROLES,
    create: ['owner', 'manager', 'cashier', 'staff'],
    pay: ['owner', 'manager', 'cashier', 'staff'],
  },
  suppliers: {
    view: ['owner', 'manager', 'stock_manager'],
    write: ['owner', 'manager', 'stock_manager'],
    delete: ['owner', 'manager', 'stock_manager'],
  },
  finance: {
    view: ['owner', 'manager'],
    create: ['owner', 'manager'],
  },
  team: {
    view: ['owner', 'manager'],
    write: ['owner', 'manager'],
    delete: ['owner', 'manager'],
  },
  audit: {
    view: ['owner', 'manager'],
  },
  settings: {
    view: ['owner', 'manager'],
    write: ['owner'],
  },
};

export function hasPermission(role: string | null | undefined, module: Module, action: string): boolean {
  // 'admin' = administrateur plateforme, hors matrice tenant — a toujours accès
  // (miroir de require_admin côté backend, séparé de cette matrice par rôle).
  const r = (role || '').trim().toLowerCase();
  if (r === 'admin') return true;
  const allowed = PERMISSIONS[module]?.[action] ?? [];
  return allowed.includes(r as Role);
}

export function decodeRoleFromToken(token: string | null): string | null {
  if (!token) return null;
  try {
    const payload = JSON.parse(atob(token.split('.')[1]));
    return payload.role || null;
  } catch {
    return null;
  }
}

/** Route (préfixe de pathname) -> permission requise pour y accéder.
 *  Utilisée pour filtrer le menu et garder l'utilisateur hors des pages
 *  non autorisées (voir layout.tsx). Routes absentes de cette liste
 *  (login, register, admin/*, whatsapp...) restent ouvertes à tout
 *  utilisateur authentifié — hors du périmètre des 14 modules à contrôler. */
export const ROUTE_PERMISSIONS: { prefix: string; module: Module; action: string }[] = [
  { prefix: '/dashboard', module: 'dashboard', action: 'view' },
  { prefix: '/analytics', module: 'dashboard', action: 'view' },
  { prefix: '/pos', module: 'orders', action: 'create' },
  { prefix: '/products', module: 'products', action: 'view' },
  { prefix: '/categories', module: 'categories', action: 'view' },
  { prefix: '/crm', module: 'clients', action: 'view' },
  { prefix: '/segments', module: 'clients', action: 'view' },
  { prefix: '/dettes', module: 'debts', action: 'view' },
  { prefix: '/sales', module: 'orders', action: 'view' },
  { prefix: '/orders', module: 'orders', action: 'view' },
  { prefix: '/suppliers', module: 'suppliers', action: 'view' },
  { prefix: '/finance', module: 'finance', action: 'view' },
  { prefix: '/team', module: 'team', action: 'view' },
  { prefix: '/audit', module: 'audit', action: 'view' },
  { prefix: '/settings', module: 'settings', action: 'view' },
  { prefix: '/billing', module: 'settings', action: 'view' },
];

/** Première route de la navigation principale accessible à ce rôle —
 *  utilisée pour rediriger hors d'une page non autorisée sans supposer
 *  que /dashboard (réservé owner/manager) convient à tout le monde. */
export function firstAllowedRoute(role: string | null | undefined): string {
  const order: { href: string; module: Module; action: string }[] = [
    { href: '/dashboard', module: 'dashboard', action: 'view' },
    { href: '/pos', module: 'orders', action: 'create' },
    { href: '/sales', module: 'orders', action: 'view' },
    { href: '/products', module: 'products', action: 'view' },
    { href: '/crm', module: 'clients', action: 'view' },
  ];
  for (const item of order) {
    if (hasPermission(role, item.module, item.action)) return item.href;
  }
  return '/settings';
}

/** Rôle courant décodé depuis le JWT stocké en localStorage — même source
 *  que layout.tsx, dupliquée volontairement ici pour rester utilisable
 *  depuis n'importe quelle page sans dépendre d'un contexte React partagé. */
export function useCurrentRole(): string | null {
  const [role, setRole] = useState<string | null>(null);
  useEffect(() => {
    setRole(decodeRoleFromToken(localStorage.getItem('boutikflow_access_token')));
  }, []);
  return role;
}

export function usePermission(module: Module, action: string): boolean {
  const role = useCurrentRole();
  return hasPermission(role, module, action);
}

File: backend/app/core/permissions.py
"""
Permissions par rôle — source unique de vérité sur ce que chaque rôle
peut faire dans chaque module.

Mirroré côté frontend (frontend/src/lib/permissions.ts) pour l'affichage
(menus, boutons), mais SEULE cette version fait autorité : le frontend ne
reflète que ce qui est de toute façon appliqué ici. Un utilisateur qui
contournerait l'interface (appel direct à l'API) reste bloqué par ces
dépendances FastAPI, jamais par le frontend seul.

Complète (ne remplace pas) app.core.deps.require_owner_or_manager, déjà
utilisé par Équipe/Audit/la création de transactions Finance — ces
modules restent inchangés, déjà corrects.
"""
from typing import Annotated, Callable
from fastapi import Depends, HTTPException, status

from app.core.deps import get_current_user, CurrentUser

Module = str
Action = str

# module -> action -> rôles autorisés (RoleEnum.value, voir
# app.modules.auth.models.RoleEnum — "staff" = Employé).
PERMISSIONS: dict[Module, dict[Action, set[str]]] = {
    "dashboard": {
        "view": {"owner", "manager"},
    },
    "products": {
        "view": {"owner", "manager", "stock_manager", "cashier", "staff"},
        "write": {"owner", "manager", "stock_manager"},
        "delete": {"owner", "manager", "stock_manager"},
    },
    "categories": {
        "view": {"owner", "manager", "stock_manager", "cashier", "staff"},
        "write": {"owner", "manager", "stock_manager"},
        "delete": {"owner", "manager", "stock_manager"},
    },
    "stock": {
        "write": {"owner", "manager", "stock_manager"},
    },
    "orders": {
        "view": {"owner", "manager", "stock_manager", "cashier", "staff"},
        "create": {"owner", "manager", "cashier", "staff"},
        "cancel": {"owner", "manager"},
    },
    "returns": {
        "create": {"owner", "manager", "stock_manager", "cashier"},
    },
    # Reprend exactement les rôles déjà utilisés aujourd'hui via
    # require_permission("orders", ...) dans debt_router.py — isolation
    # propre du module Dettes Clients, aucun changement d'accès pour
    # personne au moment de cette migration.
    "debts": {
        "view": {"owner", "manager", "stock_manager", "cashier", "staff"},
        "create": {"owner", "manager", "cashier", "staff"},
        "pay": {"owner", "manager", "cashier", "staff"},
    },
    "clients": {
        "view": {"owner", "manager", "stock_manager", "cashier", "staff"},
        "create": {"owner", "manager", "stock_manager", "cashier", "staff"},
        "edit": {"owner", "manager", "cashier"},
        "delete": {"owner", "manager"},
    },
    "suppliers": {
        "view": {"owner", "manager", "stock_manager"},
        "write": {"owner", "manager", "stock_manager"},
        "delete": {"owner", "manager", "stock_manager"},
    },
    "finance": {
        "view": {"owner", "manager"},
        "create": {"owner", "manager"},
    },
    "team": {
        "view": {"owner", "manager"},
        "write": {"owner", "manager"},
        "delete": {"owner", "manager"},
    },
    "audit": {
        "view": {"owner", "manager"},
    },
    "settings": {
        "view": {"owner", "manager"},
        "write": {"owner"},
    },
}


def has_permission(role: str | None, module: Module, action: Action) -> bool:
    role_key = (role or "").strip().lower()
    allowed = PERMISSIONS.get(module, {}).get(action, set())
    return role_key in allowed


def require_permission(module: Module, action: Action) -> Callable[..., CurrentUser]:
    """Fabrique une dépendance FastAPI exigeant la permission (module,
    action) pour le rôle courant — usage identique à
    app.core.deps.require_owner_or_manager, généralisé à toute la matrice.

    Exemple : DB = Annotated[Session, Depends(get_db)]
              User = Annotated[CurrentUser, Depends(require_permission("products", "write"))]
    """
    async def _check(
        current_user: Annotated[CurrentUser, Depends(get_current_user)],
    ) -> CurrentUser:
        if not has_permission(current_user.role, module, action):
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Votre rôle ne permet pas cette action.",
            )
        return current_user
    return _check
